### PR TITLE
Prevent duplicate Opportunity creation in Enhanced Recurring Donations

### PR DESCRIPTION
**Pull request by npsp-reedestockton**
We addressed an issue with Enhanced Recurring Donation schedule changes that could cause a duplicate Opportunity to be created.

# Critical Changes

# Changes
- We addressed an issue with Enhanced Recurring Donation schedule changes that could cause duplicate Opportunities.

# Issues Closed

# Community Ideas Delivered

# New Metadata

# Deleted Metadata

### CODE REVIEW

**Comment by npsp-reedestockton** [comment]
**lurch: attach W-040307

**Comment by npsp-reedestockton** [comment]
reference original PR #5530

**Comment by npsp-reedestockton** [comment]
@force2b This is the new PR replacing #5530.  Base is now master.

**Comment by LurchTheButler** [comment]
Tracking <a href="https://foundation.my.salesforce.com/a2x1E000001Hts6QAC" target="_blank">W-040307</a>

File: src/classes/RD2_OpportunityService.cls [hunk 1]
```diff
@@ -144,13 +144,13 @@ public with sharing class RD2_OpportunityService {
         for (npe03__Recurring_Donation__c rd : rds) {
             RD2_OpportunityMatcher.Record nextDonationDateRecord = scheduleService.getNextDonationDateRecord(rd, rd.RecurringDonationSchedules__r);
 
-            if (nextDonationDateRecord == null 
-                || !nextDonationDateRecord.isNew() 
+            if (nextDonationDateRecord == null
+                || !nextDonationDateRecord.isNew()
                 || nextDonationDateRecord.getInstallment().nextDonationDate < currentDate
-            ) {
+                ) {
```
Comment (force2b) [inline src/classes/RD2_OpportunityService.cls:150]: This should be outdented

File: src/classes/RD2_OpportunityService.cls [hunk 2]
```diff
@@ -160,9 +160,9 @@ public with sharing class RD2_OpportunityService {
         Map<Id, RD2_ScheduleService.Installment> installmentByOppId = new Map<Id, RD2_ScheduleService.Installment>();
 
         for (Opportunity opp : oppsToCreate) {
-            if (createdOppIds.contains(opp.Id) 
+            if (createdOppIds.contains(opp.Id)
                 && installmentsByRDId.containsKey(opp.npe03__Recurring_Donation__c)
-            ) {
+                ) {
```
Comment (force2b) [inline src/classes/RD2_OpportunityService.cls:165]: outdent here too

File: src/classes/RD2_OpportunityService.cls [hunk 3]
```diff
@@ -237,21 +248,49 @@ public with sharing class RD2_OpportunityService {
...
                     continue;
                 }
 
                 Opportunity opp = record.getOpportunity();
                 if (record.hasInstallment()) {
                     installmentByOppId.put(opp.Id, record.getInstallment());
+                    openOpportunityMatched = true;
+                }
+
+                if (!record.hasInstallment()){
```
Comment (force2b) [inline src/classes/RD2_OpportunityService.cls:272]:
```suggestion
                if (!record.hasInstallment()) {
```

File: src/classes/RD2_OpportunityService.cls [hunk 4]
```diff
@@ -262,6 +301,53 @@ public with sharing class RD2_OpportunityService {
...
+    /***
+    * @description Compares two installments and returns the earliest of the two
+    * @param currentInstallment Installment
+    * @param earliestInstallment Installment
+    * @return RD2_ScheduleService.Installment Earlier of the two installments
+    */
+    private RD2_ScheduleService.Installment getEarliestInstallment(RD2_ScheduleService.Installment currentInstallment, RD2_ScheduleService.Installment earliestInstallment) {
+        if (earliestInstallment == null ||
+            earliestInstallment.nextDonationDate > currentInstallment.nextDonationDate
+            ) {
```
Comment (force2b) [inline src/classes/RD2_OpportunityService.cls:313]: outdent

File: src/classes/RD2_OpportunityService.cls [hunk 5]
```diff
@@ -262,6 +301,53 @@ public with sharing class RD2_OpportunityService {
...
+    /***
+    * @description Compares two Opportunities and returns the earliest of the two
+    * @param currentOpportunity Opportunity
+    * @param earliestUnmatchedOpportunity Opportunity
+    * @return Opportunity Earlier of the two Opportunities
+    */
+    private Opportunity getEarliestOpportunity(Opportunity currentOpportunity, Opportunity earliestUnmatchedOpportunity) {
+        if (earliestUnmatchedOpportunity == null ||
+            earliestUnmatchedOpportunity.CloseDate > currentOpportunity.CloseDate
+            ) {
```
Comment (force2b) [inline src/classes/RD2_OpportunityService.cls:328]: outdent

File: src/classes/RD2_OpportunityService.cls [hunk 6]
```diff
@@ -304,7 +390,7 @@ public with sharing class RD2_OpportunityService {
         if (installment.campaignId != null
             && opp.CampaignId != installment.campaignId
             && rdSettings.shouldSetInstallmentCampaign
-        ) {
+            ) {
```
Comment (force2b) [inline src/classes/RD2_OpportunityService.cls:393]: outdent

File: src/classes/RD2_OpportunityService.cls [hunk 7]
```diff
@@ -387,7 +473,7 @@ public with sharing class RD2_OpportunityService {
             if (installment != null
                 && String.isNotBlank(installment.paymentMethod)
                 && payment.npe01__Payment_Method__c != installment.paymentMethod
-            ) {
+                ) {
```
Comment (force2b) [inline src/classes/RD2_OpportunityService.cls:476]: outdent

File: src/classes/RD2_RecurringDonations_TDTM.cls [hunk 8]
```diff
@@ -145,7 +145,7 @@ public class RD2_RecurringDonations_TDTM extends TDTM_Runnable {
     ) {
         if (triggerAction != TDTM_Runnable.Action.BeforeInsert
             && triggerAction != TDTM_Runnable.Action.BeforeUpdate
-        ) {
+            ) {
```
Comment (force2b) [inline src/classes/RD2_RecurringDonations_TDTM.cls:148]: outdent here and below

**Comment by npsp-reedestockton** [comment]
@force2b PR feedback has been addressed.  Ready for your review.  Thanks!

**Comment by npsp-reedestockton** [comment]
@ravensisco This item is ready for QA now.  Thanks!

**Pull request merged**